perception/traffic_light_arbiter/src/signal_match_validator.cpp

```diff
@@ -0,0 +1,274 @@
+
+  return signal_id_set;
+}
+
+// Returns the signal with the highest confidence elements, considering a external priority
+TrafficSignal get_highest_confidence_signal(
+  const std::optional<TrafficSignal> & perception_signal,
+  const std::optional<TrafficSignal> & external_signal, const bool external_priority)
+{
+  // Returns the existing signal if only one of them exists
+  if (!perception_signal) {
+    return *external_signal;
+  }
+  if (!external_signal) {
+    return *perception_signal;
+  }
+
+  // Gives priority to the external signal if external_priority is true
+  if (external_priority) {
+    return *external_signal;
+  }
+
+  // Compiles elements into a map by shape, to compare their confidences
+  using Key = Element::_shape_type;
+  std::map<Key, std::vector<Element>> shape_element_map;
+  for (const auto & element : perception_signal->elements) {
+    shape_element_map[element.shape].emplace_back(element);
+  }
+  for (const auto & element : external_signal->elements) {
+    shape_element_map[element.shape].emplace_back(element);
+  }
+
+  TrafficSignal highest_confidence_signal;
+
+  // Assumes that both signals have the same traffic_signal_id
+  highest_confidence_signal.traffic_signal_id = perception_signal->traffic_signal_id;
+
+  // For each shape, finds the element with the highest confidence and adds it to the signal
+  for (const auto & [shape, elements] : shape_element_map) {
+    const auto highest_confidence_element = std::max_element(
+      elements.begin(), elements.end(),
+      [](const Element & a, const Element & b) { return a.confidence < b.confidence; });
+    highest_confidence_signal.elements.emplace_back(*highest_confidence_element);
+  }
+
+  return highest_confidence_signal;
+}
+
+// Determines the newer of two Time stamps
+Time get_newer_stamp(const Time & stamp1, const Time & stamp2)
+{
+  // Returns stamp1 if it is newer than stamp2, otherwise returns stamp2
+  if (stamp1.sec > stamp2.sec || (stamp1.sec == stamp2.sec && stamp1.nanosec > stamp2.nanosec)) {
+    return stamp1;
+  } else {
+    return stamp2;
+  }
+}
+
+}  // namespace util
+
+autoware_perception_msgs::msg::TrafficSignalArray SignalMatchValidator::validateSignals(
+  const TrafficSignalArray & perception_signals, const TrafficSignalArray & external_signals)
+{
+  TrafficSignalArray validated_signals;
+
+  // Set newer stamp
+  validated_signals.stamp = util::get_newer_stamp(perception_signals.stamp, external_signals.stamp);
+
+  // Create the unique set of the received id,
+  // then compare the signal element for each received signal id
+  const auto received_signal_id_set =
+    util::create_signal_id_set(perception_signals.signals, external_signals.signals);
+
+  for (const auto & signal_id : received_signal_id_set) {
+    const auto perception_result = util::find_signal_by_id(perception_signals, signal_id);
+    const auto external_result = util::find_signal_by_id(external_signals, signal_id);
+
+    // Both results doesn't exist
```

Comment:
Neither result exists

Reply:
Addressed in https://github.com/autowarefoundation/autoware.universe/pull/6423/commits/454f0490924f3d616c497aa77f5fcd038808bea6